components/script/dom/htmlimageelement.rs

```diff
@@ -871,6 +871,166 @@ impl HTMLImageElement {
+        };
+
+        let same_selected_pixel_density = match selected_pixel_density {
+            Some(den) => self.current_request.borrow().current_pixel_density.map_or(false, |density| density == den),
+            _ => self.current_request.borrow().current_pixel_density.map_or(true, |_den| false),
+        };
+
+        if same_source && same_selected_pixel_density {
+            return;
+        }
+
+        // Step 6
+        if let Ok(img_url) = base_url.join(&String::from(src.clone())) {
+            // Step 12
+            *self.pending_request.borrow_mut() = ImageRequest {
+                state: State::Unavailable,
+                parsed_url: Some(img_url.clone()),
+                source_url: None,
+                image: None,
+                metadata: None,
+                blocker: None,
+                final_url: None,
+                current_pixel_density: None,
+            };
+
+            let window = window_from_node(self);
+            let image_cache = window.image_cache();
+            // Step 14
+            let response =
+                image_cache.find_image_or_metadata(img_url.clone().into(),
+                                                   UsePlaceholder::No,
+                                                   CanRequestImages::No);
+            if let Ok(ImageOrMetadataAvailable::ImageAvailable(_image, _url)) = response {} else {
+                let window = window_from_node(self);
+                let image_cache = window.image_cache();
+                // Step 14.1
+                let response =
+                    image_cache.find_image_or_metadata(img_url.clone().into(),
+                                                       UsePlaceholder::Yes,
+                                                       CanRequestImages::Yes);
+                match response {
+                    // TODO Handle multipart/x-mixtape
+                    Ok(ImageOrMetadataAvailable::ImageAvailable(_image, _url)) => {}
+                    Ok(ImageOrMetadataAvailable::MetadataAvailable(_m)) => { }
+                    Err(ImageState::Pending(_id)) => { }
+                    Err(ImageState::NotRequested(_id)) => { }
+                    Err(ImageState::LoadError) => {
+                        *self.pending_request.borrow_mut() = ImageRequest {
+                            state: State::Unavailable,
+                            parsed_url: None,
+                            source_url: None,
+                            image: None,
+                            metadata: None,
+                            blocker: None,
+                            final_url: None,
+                            current_pixel_density: None,
+                        };
+                    }
+
+                }
+            }
+
+            let this = Trusted::new(self);
+            let new_selected_source = src.to_string();
+            let relevant_mutation = self.generation.get() != *generation;
+            // Step 15
```

Comment:
Step 15 belongs in a `finish_reacting_to_environment_change` method that we can call at the appropriate time.